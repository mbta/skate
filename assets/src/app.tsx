// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
