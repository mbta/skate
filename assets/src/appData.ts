const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
