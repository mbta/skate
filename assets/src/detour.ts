import { LatLngLiteral } from "leaflet"
import { Shape, ShapePoint } from "./schedule"

export interface DetourShape {
  coordinates: ShapePoint[]
  directions: {
    instruction: string
  }[]
}

export interface OriginalRoute {
  routeName: string
  routeDescription: string
  routeOrigin: string
  routeDirection: string
  shape: Shape
  center: LatLngLiteral
  zoom: number
}
