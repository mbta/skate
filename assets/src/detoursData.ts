import { array, Infer, number, string, type } from "superstruct"
import { DraftDetours } from "./models/detours"

export const DraftDetoursData = array(
  type({
    route: string(),
    direction: string(),
    name: string(),
    intersection: string(),
    active_since: number(),
  })
)
export type DraftDetoursData = Infer<typeof DraftDetoursData>

export const draftDetoursFromData = (
  draftDetoursData: DraftDetoursData
): DraftDetours =>
  draftDetoursData.map((draft) => ({
    route: draft.route,
    direction: draft.direction,
    name: draft.name,
    intersection: draft.intersection,
    activeSince: draft.active_since,
  }))
