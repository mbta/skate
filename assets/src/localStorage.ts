export const saveState = (stateKey: string, state: object) => {
  const serializedState = JSON.stringify(state)
  localStorage.setItem(stateKey, serializedState)
}

export const loadState = (stateKey: string): object | undefined => {
  const serializedState = localStorage.getItem(stateKey)

  if (serializedState === null) {
    return undefined
  }

  return JSON.parse(serializedState)
}
