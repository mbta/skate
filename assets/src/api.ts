import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
