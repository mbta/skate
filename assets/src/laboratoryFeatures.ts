import appData from "./appData"

const featureIsEnabled = (key: string): boolean => {
  const data = appData()

  if (data === undefined || data.laboratoryFeatures === undefined) {
    return false
  }

  const features = JSON.parse(data.laboratoryFeatures)
  if (features[key] === undefined) {
    return false
  }

  return features[key]
}

export default featureIsEnabled
