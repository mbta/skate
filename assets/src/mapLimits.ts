import { create, Infer, number, type } from "superstruct"
import appData from "./appData"

const MapLimits = type({
  north: number(),
  south: number(),
  east: number(),
  west: number(),
})
type MapLimits = Infer<typeof MapLimits>

const getMapLimits = (): MapLimits | null => {
  const mapLimitsJson = appData()?.mapLimits

  if (mapLimitsJson === undefined) {
    return null
  }

  const mapLimits = create(JSON.parse(mapLimitsJson) as unknown, MapLimits)

  return mapLimits
}

export default getMapLimits
