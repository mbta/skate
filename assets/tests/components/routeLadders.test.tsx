import React from "react"
import renderer from "react-test-renderer"
import RouteLadders from "../../src/components/routeLadders"
import { Route, TimepointsByRouteId } from "../../src/skate"

test("renders a route ladder", () => {
  const routes: Route[] = [{ id: "1" }, { id: "28" }]
  const timepointsByRouteId: TimepointsByRouteId = {
    "1": [{ id: "WASMA" }, { id: "MELWA" }, { id: "HHGAT" }],
    "28": [{ id: "MATPN" }, { id: "WELLH" }, { id: "MORTN" }],
    "71": undefined,
    "73": null,
  }

  const tree = renderer
    .create(
      <RouteLadders
        routes={routes}
        timepointsByRouteId={timepointsByRouteId}
        vehiclesByRouteId={{}}
        selectedVehicleId={undefined}
      />
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
