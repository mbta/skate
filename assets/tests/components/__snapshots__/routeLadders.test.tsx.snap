// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a route ladder 1`] = `
<div
  className="m-route-ladders"
>
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      1
    </div>
  </div>
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>
  <div
    className="m-layover-box m-layover-box--top"
  />
  <div
    className="m-ladder"
    style={
      Object {
        "width": 184,
      }
    }
  >
    <svg
      className="m-ladder__svg"
      height={0}
      viewBox="-92 -20 184 0"
      width={184}
    >
      <line
        className="m-ladder__line"
        x1={-40}
        x2={-40}
        y1="0"
        y2={-40}
      />
      <line
        className="m-ladder__line"
        x1={40}
        x2={40}
        y1="0"
        y2={-40}
      />
      <g
        className="m-ladder__headway-lines"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-0}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-0}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="WASMA Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-0}
      >
        WASMA
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-20}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-20}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MELWA Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-20}
      >
        MELWA
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-40}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-40}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="HHGAT Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-40}
      >
        HHGAT
      </text>
    </svg>
    <div
      className="__react_component_tooltip place-top type-dark"
      data-id="tooltip"
    />
  </div>
  <div
    className="m-layover-box m-layover-box--bottom"
  />
  <div
    className="m-incoming-box"
  />
  <div
    className="m-route-ladder__header"
  >
    <button
      className="m-close-button"
      onClick={[Function]}
    >
      <span
        className=""
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </button>
    <div
      className="m-route-ladder__route-name"
    >
      28
    </div>
  </div>
  <div
    className="m-route-ladder__controls"
  >
    <button
      className="m-route-ladder__reverse"
      onClick={[Function]}
    >
      <span
        className="m-route-ladder__reverse-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Reverse
    </button>
  </div>
  <div
    className="m-layover-box m-layover-box--top"
  />
  <div
    className="m-ladder"
    style={
      Object {
        "width": 184,
      }
    }
  >
    <svg
      className="m-ladder__svg"
      height={0}
      viewBox="-92 -20 184 0"
      width={184}
    >
      <line
        className="m-ladder__line"
        x1={-40}
        x2={-40}
        y1="0"
        y2={-40}
      />
      <line
        className="m-ladder__line"
        x1={40}
        x2={40}
        y1="0"
        y2={-40}
      />
      <g
        className="m-ladder__headway-lines"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-0}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-0}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MATPN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-0}
      >
        MATPN
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-20}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-20}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="WELLH Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-20}
      >
        WELLH
      </text>
      <circle
        className="m-ladder__stop-circle"
        cx={-40}
        cy={-40}
        r="3"
      />
      <circle
        className="m-ladder__stop-circle"
        cx={40}
        cy={-40}
        r="3"
      />
      <text
        className="m-ladder__timepoint-name"
        data-event="click"
        data-tip="MORTN Name"
        dominantBaseline="middle"
        textAnchor="middle"
        x="0"
        y={-40}
      >
        MORTN
      </text>
    </svg>
    <div
      className="__react_component_tooltip place-top type-dark"
      data-id="tooltip"
    />
  </div>
  <div
    className="m-layover-box m-layover-box--bottom"
  />
  <div
    className="m-incoming-box"
  />
</div>
`;
